In [1]:
import os
import sys
module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)
module_path = os.path.abspath(os.path.join('../..'))
if module_path not in sys.path:
    sys.path.append(module_path)
    
from src import gen_spectra

# Fancy tiling algorithm
Inputs:
1. 
2. 
3. 

Outputs: 



Oracles: 
1. Mass of any AA for any charge
2. Precursor mass of any AA sequence

## Edge cases


In [60]:
sorted(gen_spectra.gen_spectrum('DLQTLALWSRM')['spectrum'])

[58.520747935,
 75.53280128499999,
 115.06277993500001,
 116.034219435,
 150.058326135,
 153.58335678499998,
 179.09206893500001,
 197.09937078499996,
 229.118283435,
 229.61590843500002,
 286.157940435,
 290.139027285,
 306.15943713499996,
 321.676497435,
 346.681059285,
 357.176861435,
 378.218529435,
 382.199616285,
 393.19146513499993,
 438.741648285,
 458.224540435,
 471.25818593500003,
 489.265487785,
 514.7741999350001,
 553.2947767850001,
 571.308604435,
 579.270778135,
 592.824755435,
 609.836808785,
 642.345718435,
 658.344997935,
 667.350280285,
 692.354842135,
 755.429782435,
 763.391956135,
 876.476020135,
 941.5090954350001,
 977.523699135,
 1028.5411234350001,
 1105.582277135,
 1184.642234435,
 1218.6663411349998,
 1315.682719435,
 1333.6932841349999]

In [13]:
observed = [84.04356384277344, 86.095703125, 101.07088470458984, 201.12228393554688, 202.1270751953125, 212.10423278808594, 229.1165008544922, 246.14280700683594, 306.16021728515625, 357.17578125, 393.1902770996094, 430.2205505371094, 440.2127685546875, 458.2236022949219, 553.2975463867188, 579.2705078125, 624.333251953125, 642.3447875976562, 658.3429565429688, 692.3502197265625, 763.3924560546875, 764.3958129882812, 876.477783203125, 977.5200805664062, 978.5184326171875]

In [47]:
prefix = 'WSRM'
y1 = gen_spectra.gen_spectrum(prefix, ion='y', charge=1)['spectrum'][-1]
b1 = gen_spectra.gen_spectrum(prefix, ion='b', charge=1)['spectrum'][-1]
y2 = gen_spectra.gen_spectrum(prefix, ion='y', charge=2)['spectrum'][-1]
b2 = gen_spectra.gen_spectrum(prefix, ion='b', charge=2)['spectrum'][-1]

In [63]:
sub_y1 = [x - y1 for x in observed if x - y1 > 0 and x < 980]
sub_b1 = [x - b1 for x in observed if x - b1 > 0 and x < 980]
sub_y2 = [x - y2 for x in observed if x - y2 > 0 and x < 980]
sub_b2 = [x - b2 for x in observed if x - b2 > 0 and x < 980]

In [64]:
sub_y1

[45.062473818125,
 63.07400946265625,
 79.07217840796875,
 113.0794415915625,
 184.1216779196875,
 185.12503485328125,
 297.207005068125,
 398.24930243140625,
 399.2476544821875]

In [66]:
sub_y2

[16.021190000156253,
 67.036753965,
 103.05124981460938,
 140.08152325210938,
 150.0737412696875,
 168.08457500992188,
 263.15851910171875,
 289.1314805275,
 334.194224668125,
 352.20576031265625,
 368.20392925796875,
 402.2111924415625,
 473.2534287696875,
 474.25678570328125,
 586.338755918125,
 687.3810532814063,
 688.3794053321875]

In [67]:
# sub_b1

[18.01029437750003,
 63.07303851812503,
 81.08457416265628,
 97.08274310796878,
 131.09000629156253,
 202.13224261968753,
 203.13559955328128,
 315.21756976812503,
 416.2598671314063,
 417.25821918218753]

In [82]:
# sub_b2

In [80]:
gen_spectra.gen_spectrum('AL', ion='y')['spectrum']

[132.101905135, 203.139019135, 66.554590785, 102.073147785]

In [71]:
from src.scoring import scoring

In [100]:
scoring.score_sequence(observed, sorted(gen_spectra.gen_spectrum('DLQTLAL', ion='b')['spectrum']), 20)

5

In [83]:
gen_spectra.gen_spectrum('ALWSRM', ion='y', charge=1)['spectrum'][-1]

763.391956135

In [84]:
gen_spectra.gen_spectrum('ALWSRM', ion='y', charge=2)['spectrum'][-1]

382.199616285

In [85]:
observed

[84.04356384277344,
 86.095703125,
 101.07088470458984,
 201.12228393554688,
 202.1270751953125,
 212.10423278808594,
 229.1165008544922,
 246.14280700683594,
 306.16021728515625,
 357.17578125,
 393.1902770996094,
 430.2205505371094,
 440.2127685546875,
 458.2236022949219,
 553.2975463867188,
 579.2705078125,
 624.333251953125,
 642.3447875976562,
 658.3429565429688,
 692.3502197265625,
 763.3924560546875,
 764.3958129882812,
 876.477783203125,
 977.5200805664062,
 978.5184326171875]

In [86]:
763.391956135 - gen_spectra.gen_spectrum('WSRM', ion='y', charge=1)['spectrum'][-1]

184.121178

1. Find the first y ion that is a product of a hybrid
    1. This is the first y ion to the left of the hybrid junction
        * `AL-WSRM`, we found the mass of `ALWSRM` in observed
2. Find the y1 and y2 masses of `ALWSRM`
3. Subtract the mass of `WSRM` from `ALWSRM`
    1. gives us 184 for the y1 case
    2. Found by just checking the cases (theoretical spectra) for singly and doubly charged
4. Check if this mass was found in our y1_sub set
    1. y1_sub was any observed mass - y1 of `WSRM` > 0
    

In [ ]:
'DQLTLAL'
         'SRM'

'DLQTLAL-WS'
     'AL-WSRM'
    
'DQLTLAL-WSRM'